core/platform/auth/firebase_auth_services.py

```diff
@@ -0,0 +1,240 @@
+
+    @staticmethod
+    def get_model_association_to_user():
+        """Currently, the model holds identifiers relevant only for backend that
+        should not be exported.
+        """
+        return base_model.MODEL_ASSOCIATION_TO_USER.NOT_CORRESPONDING_TO_USER
+
+    @classmethod
+    def get_export_policy(cls):
+        """Model doesn't contain any data directly corresponding to a user.
+        Currently, the model holds authentication details relevant only for
+        backend, and no exportable user data. It may contain user data in
+        the future.
+        """
+        return dict(
+            super(_UserIdByFirebaseSubModel, cls).get_export_policy(),
+            **{'user_id': base_model.EXPORT_POLICY.NOT_APPLICABLE})
+
+    @classmethod
+    def apply_deletion_policy(cls, user_id):
+        """Delete instances of _UserIdByFirebaseSubModel for the user.
+
+        Args:
+            user_id: str. The ID of the user whose data should be deleted.
+        """
+        datastore_services.delete_multi(
+            cls.query(cls.user_id == user_id).fetch(keys_only=True))
+
+    @classmethod
+    def has_reference_to_user_id(cls, user_id):
+        """Check whether _UserIdByFirebaseSubModel exists for the given user.
+
+        Args:
+            user_id: str. The ID of the user whose data should be checked.
+
+        Returns:
+            bool. Whether any _UserIdByFirebaseSubModel refers to the given user
+            ID.
+        """
+        return cls.query(cls.user_id == user_id).get(keys_only=True) is not None
+
+    @classmethod
+    def get_by_user_id(cls, user_id):
+        """Fetch a entry by user ID.
+
+        Args:
+            user_id: str. The user ID.
+
+        Returns:
+            _UserIdByFirebaseSubModel. The model with user_id field equal to
+            user_id argument.
+        """
+        return cls.query(cls.user_id == user_id).get()
+
+
+def get_user_id_from_sub(sub):
+    """Returns the user ID associated to the given subject identifier.
+
+    Args:
+        sub: str. The subject identifier.
+
+    Returns:
+        str|None. The user ID associated to the given subject, or None if no
+        association exists yet.
+    """
+    mapping = _UserIdByFirebaseSubModel.get_by_id(sub)
+    return None if mapping is None else mapping.user_id
+
+
+def get_multi_user_ids_from_subs(subs):
+    """Returns the user IDs associated to the given subject identifiers.
+
+    Args:
+        subs: list(str). The subject identifiers.
+
+    Returns:
+        list(str|None). The user ID associated to each of the given subject,
+        or None for associations that don't exist.
+    """
+    return [None if mapping is None else mapping.user_id
+            for mapping in _UserIdByFirebaseSubModel.get_multi(subs)]
+
+
+def associate_sub_to_user_id(sub, user_id):
+    """Commits the association between user ID and Firebase subject identifier.
+
+    Args:
+        sub: str. The subject identifier of the user.
```

Comment:
How do you feel about calling this firebase_subject_id and similar throughout? I am a bit worried that "sub" is a rather generic prefix and could refer to a lot of things (subscriptions, etc.)

Reply:
Ditto, done.